wallet-new/src/Cardano/Wallet/Kernel.hs

```diff
@@ -149,6 +155,25 @@ applyBlocks :: PassiveWallet
             -> IO ()
 applyBlocks = mapM_ . applyBlock
 
+-- | Switch to a new fork
+--
+-- NOTE: The Ouroboros protocol says that this is only valid if the number of
+-- resolved blocks exceeds the length of blocks to roll back.
+switchToFork :: PassiveWallet
+             -> Int             -- ^ Number of blocks to roll back
+             -> [ResolvedBlock] -- ^ Blocks in the new fork
+             -> IO ()
+switchToFork pw@PassiveWallet{..} n bs = do
+    blockssByAccount <- mapM (prefilterBlock' pw) bs
```

Comment:
Something we might need to revisit later (jotting down out loud notes here): another good testament for CBR-341, as `prefilterBlock'` is calling `Keystore.toList` every time, and that might be expensive if the number of blocks to resolve is big.

Reply:
Added a [comment to CBR-341](https://iohk.myjetbrains.com/youtrack/issue/CBR-341#comment=93-26006) that this `toList` should be removed completely